### PR TITLE
Removed an unsafe usage of `EntityMinecart.getType()`

### PR DESCRIPTION
**Pull request by liach**
Here, I redirected the call to a Forge provided methods instead. Don't worry too much about the wrong branch name.

### CODE REVIEW

**Comment by mezz** [comment]
Hmm `canBeRidden()` is pretty ancient, I wonder if it's needed any more. I looked at the history and it predates `getType()`, the original used `this instanceof ItemRideableMinecart`.

Anyway ignoring that, this PR looks good to me.

**Comment by liach** [comment]
@mezz As an enthusiast of Railcraft, here I believe that `canBeRidden()` is definitely needed.

**Comment by mezz** [comment]
@liach Could you fix a related problem here too?
`EntityMinecart` calls `BlockRailBase.getShapeProperty()` in two places where it should call `BlockRailBase.getRailDirection` instead.

File: patches/minecraft/net/minecraft/block/BlockRailBase.java.patch [hunk 1]
```diff
@@ -48,8 +61,20 @@
...
+-        if (((BlockRailBase.EnumRailDirection)p_180663_3_.func_177229_b(this.func_176560_l())).func_177018_c())
++        if (getRailDirection(p_180663_1_, p_180663_2_, p_180663_3_, null).func_177018_c())
+         {
+             p_180663_1_.func_175685_c(p_180663_2_.func_177984_a(), this, false);
+         }
+@@ -160,8 +160,101 @@
+         }
+     }
  
++    //Forge: Use getRailDirection(IBlockAccess, BlockPos, IBlockState, EntityMinecart) for enhanced ability
```
Comment (liach) [inline patches/minecraft/net/minecraft/block/BlockRailBase.java.patch:77]: @mezz Should I deprecate this? If yes, then we need to add a method to cycle rail direction property.

**Comment by liach** [comment]
I replaced such calls in `EntityMinecart`, `BlockRailBase`, and `ItemMinecart` classes.

File: patches/minecraft/net/minecraft/block/BlockRailBase.java.patch (hunk 1 above)
Comment (mezz) [reply patches/minecraft/net/minecraft/block/BlockRailBase.java.patch:77]: No, there are legitimate uses for it like you said. I think this comment is good enough.

**Comment by mezz** [comment]
Looks good, thanks @liach.

**Comment by liach** [comment]
@mezz I also tested travis ci, but not by reopening but by rebasing and force pushing.

**Comment by mezz** [comment]
Any update should trigger it. 
I was testing with closing/opening because I want to run it on other people's PRs.

**Pull request merged**